loris/jp2_extractor.py

```diff
@@ -14,31 +23,95 @@ class JP2ExtractionError(LorisException):
     pass
 
 
-class JP2ExtractorMixin(object):
+def _parse_length(jp2, box_name):
     """
-    Contains logic for parsing a JPEG2000 images.
+    Internally, a JP2 is a series of boxes.  Within each box,
+    the first 4 bytes are a length field, measuring the size of the box
+    (including the length field itself).
+
+    If ``jp2`` is at the start of a box, return the length of the next box.
+
+    See § I.4.
 
-    We don't use Pillow for JPEG2000, because it doesn't expose advanced
-    features such as tiles or color profiles.
+    """
+    length_field = jp2.read(4)
+    try:
+        return struct.unpack('>I', length_field)[0]
+    except struct.error as err:
+        raise JP2ExtractionError(
+            "Error reading the length field in the %s box: %r" %
+            (box_name, err)
+        )
+
+
+class JP2Extractor(object):
+    """
+    Contains logic for parsing a JPEG2000 images.
 
     This class is meant to be used as a mixin on ImageInfo, but is kept
     separately for easier testing.
     """
     __slots__ = ()
 
+    def _check_signature_box(self, jp2):
+        """
+        The first 12 bytes of a JP2 file are the "JPEG 2000 Signature box".
+        Quoting ISO/IEC 15444-1:2000(E) § I.5.1:
+
+            For file verification purposes, this box can be considered a
+            fixed-length 12-byte string which shall have the value:
+            0x0000 000C 6A50 2020 0D0A 870A.
+
+        """
+        signature = jp2.read(12)
+        if signature != b'\x00\x00\x00\x0c\x6a\x50\x20\x20\x0d\x0a\x87\x0a':
+            raise JP2ExtractionError("Bad signature box: %r" % signature)
+
+    def _check_file_type_box(self, jp2):
+        """
+        After the Signature box is the "File Type box" (see § I.5.2).
+        This is a variable length box, laid out as follows:
+
+            0 - 3   Length
+            4 - 7   Type, which must be 'ftyp' (0x6674 7970)
+            8 - 11  Brand, the only allowed value of which is 'jp2\040'
+            12+     Minor version of the JP2 spec; compatibility list for other
+                    standards.  This is a variable length field which we don't
+                    need to worry about.
+
+        """
+        # We should probably check the length is valid; in practice we don't
+        # right now.
+        file_type_box_length = _parse_length(jp2, 'File Type')
+
+        file_type = jp2.read(4)
+        if file_type != b'ftyp':
+            raise JP2ExtractionError(
+                "Bad type in the File Type box: %r" % file_type
+            )
+
+        file_brand = jp2.read(4)
+        if file_brand != b'jp2\040':
+            raise JP2ExtractionError(
+                "Bad brand in the File Type box: %r" % file_brand
+            )
+
+        # We've already consumed 12 bytes of the box reading the length, type,
+        # and brand fields.  Consume and discard the remaining bytes.
+        jp2.read(file_type_box_length - 12)
```

Comment:
This is the one deviation from the previous code – if you look at the next block in `extract_jp2()`, it keeps reading a new byte until it finds the start of the `ihdr` box. Doing it this way allows us to jump straight to the start of the next box.

Note that this may fail if the length is 0 (see § I.4):

> If the value of [the LBox] is 0, then the length of the box was not known when the LBox field was written.

so I may keep the current code that searches for `ihdr`; but it feels unlikely to me that this would be 0 in practice.

Reply:
Could you add a check that file_type_box_length - 12 is greater than 0? If not, we could just not do that read at all. Would that work?

Reply:
Yeah, I’ve realised this falls over if you call `read()` with a negative number.

So I think I want a compromise: if the length is present, we just use that, otherwise we use the current method to find `ihdr`, then back up 4 bytes. Either way, we end at the beginning of the next box.

Reply:
OK.